asm/kanread.s

```diff
@@ -0,0 +1,209 @@
+    li      $at, 0x8800
+    slt     $at, $a0, $at
+    bnez    $at, .nonkanji
+     // 0xBC is number of glyphs in one block in the `kanji` file:
+     // 0x100 possible codepoints with the same byte1
+     // - 0x40 unused at beginning 
+     // - 1 unused at 0x7F 
+     // - 3 unused at 0xFD, 0xFE, 0xFF
+     li     $a2, 0xBC
+    // Get byte1 and adjust so starts at 0
+    srl     $a1, $a0, 8
+    addi    $a1, $a1, -0x88
+    multu   $a2, $a1
+    // Get byte2 and adjust so starts at 0
+    andi    $a3, $a0, 0xff
+    addi    $a3, $a3, -0x40
+    slti    $at, $a3, 0x40
+    mflo    $a2
+    // 0x__7F is always empty and elided in the file, so decrement
+    bnezl   $at, .kanji_lower_halfblock
+     mflo   $a2
+    addi    $a3, $a3, -1
+    mflo    $a2
+.kanji_lower_halfblock:
+    addi    $a3, $a3, 0x30A
+    add     $a3, $a3, $a2
+    jr      $ra
+     sll    $v0, $a3, 7
+     // returns (0x30A + (adjusted byte2) + (adjusted byte1) * 0xBC) * FONT_CHAR_TEX_SIZE
+
+// Non-kanji are arranged with irregular gaps, use the lookup table.
+.nonkanji:
+    // Get byte1 and adjust so starts at 0
+    srl     $a1, $a0, 8
+    addi    $a1, $a1, -0x81
+    multu   $a2, $a1
+    // Get byte2 and adjust so starts at 0
+    andi    $a3, $a0, 0xFF
+    addi    $a3, $a3, -0x40
+    slti    $at, $a3, 0x40
+    mflo    $a2
+    // 0x__7F is always empty and elided in the file, so decrement
+    bnezl   $at, .nonkanji_lower_halfblock
+     mflo   $a2
+    addi    $a3, $a3, -1
+    mflo    $a2
+.nonkanji_lower_halfblock:
+    add     $a3, $a3, $a2
+    lui     $a2, %hi(sKanjiNonKanjiIndices)
+    sll     $a3, $a3, 1
+    addiu   $a2, %lo(sKanjiNonKanjiIndices)
+    add     $a3, $a3, $a2
+    lh      $a2, ($a3)
+    jr      $ra
+     sll    $v0, $a2, 7
+     // returns sKanjiNonKanjiIndices[(adjusted byte2) + (adjusted byte1) * 0xBC] * FONT_CHAR_TEX_SIZE
+END(Kanji_OffsetFromShiftJIS)
+
+// Nearly equivalent C code (Equivalent for all valid input, will behave differently on overflow due to the use of `add` in the original):
+// s32 Kanji_OffsetFromShiftJIS(s32 shiftJISCodepoint) {
+//     s32 byte1;
+//     s32 byte2;
+//
+//     if (shiftJISCodepoint >= 0x8800) {
+//         byte1 = shiftJISCodepoint >> 8;
+//         byte1 -= 0x88;
+//
+//         byte2 = (shiftJISCodepoint & 0xFF);
+//         byte2 -= 0x40;
+//         if (byte2 >= 0x40) {
+//             byte2--;
+//         }
+//
+//         return (0x30A + byte2 + byte1 * 0xBC) * 0x80;
+//     } else {
+//         byte1 = shiftJISCodepoint >> 8;
+//         byte1 -= 0x81;
+//
+//         byte2 = (shiftJISCodepoint & 0xFF);
+//         byte2 -= 0x40;
+//         if (byte2 >= 0x40) {
+//             byte2--;
+//         }
+//
+//         return sKanjiNonKanjiIndices[byte2 + byte1 * 0xBC] * 0x80;
+//     }
+// }
+
+DATA(sKanjiNonKanjiIndices)
+//                 0x___0  0x___1  0x___2  0x___3  0x___4  0x___5  0x___6  0x___7  0x___8  0x___9  0x___A  0x___B  0x___C  0x___D  0x___E  0x___F
```

Comment:
I would add a comment about the (good) way this is formatted, something along the lines of
```suggestion

// Formatting follows the mapping from Shift-JIS codepoints,
// and the data is divided into blocks by high byte of the codepoint
DATA(sKanjiNonKanjiIndices)
//                 0x___0  0x___1  0x___2  0x___3  0x___4  0x___5  0x___6  0x___7  0x___8  0x___9  0x___A  0x___B  0x___C  0x___D  0x___E  0x___F
```
I'm not convinced by my own comment but something like that